src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java

```diff
@@ -160,6 +162,83 @@ public void testPrevAndNextTest(final GATKRead read, final CigarOperator firstOp
         Assert.assertEquals(pe2.getNextOnGenomeCigarElement(), null);
     }
 
+    @Test(dataProvider = "PrevAndNextTest")
+    public void testImmediateBeforeAndAfterTest(final GATKRead read, final CigarOperator firstOp, final CigarOperator lastOp, final List<CigarOperator> ops) {
+        final AlignmentStateMachine state = new AlignmentStateMachine(read);
+
+        //before the first 'op' from the list
+        state.stepForwardOnGenome();
+        final PileupElement pe1 = state.makePileupElement();
+        Assert.assertEquals(pe1.getAdjacentOperator(PileupElement.Direction.PREV), null);
+        Assert.assertEquals(pe1.getAdjacentOperator(PileupElement.Direction.NEXT), ops.get(0));
+        for (final CigarOperator op : CigarOperator.values()){
+            Assert.assertEquals(pe1.isImmediatelyBefore(op), ops.get(0) == op, op.toString());
+            Assert.assertFalse(pe1.isImmediatelyAfter(op), op.toString());
+        }
+
+        //after the first 'op' from the list
+        state.stepForwardOnGenome();
+        final PileupElement pe2 = state.makePileupElement();
+        Assert.assertEquals(pe2.getAdjacentOperator(PileupElement.Direction.PREV), ops.get(ops.size()-1));
+        Assert.assertEquals(pe2.getAdjacentOperator(PileupElement.Direction.NEXT), null);
+        for (final CigarOperator op : CigarOperator.values()){
+            Assert.assertFalse(pe2.isImmediatelyBefore(op), "immediately before " + op);
+            Assert.assertEquals(pe2.isImmediatelyAfter(op), op ==  ops.get(ops.size()-1), "immediately after " + op);
+        }
+    }
+
+    @DataProvider(name = "PrevAndNextTest_simple")
+    public Object[][] makePrevAndNextTest_simple() {
+        final List<Object[]> tests = new LinkedList<>();
+
+        for ( final CigarOperator firstOp : Arrays.asList(CigarOperator.M) ) {
+            for ( final CigarOperator lastOp : Arrays.asList(CigarOperator.M) ) {
+                for ( final CigarOperator middleOp : Arrays.asList(CigarOperator.I, CigarOperator.P, CigarOperator.S) ) {
+                    final int readLength = 3;
+                    final GATKRead read = ArtificialReadUtils.createArtificialRead(header, "read", 0, 1, readLength);
+                    read.setBases(Utils.dupBytes((byte) 'A', readLength));
+                    read.setBaseQualities(Utils.dupBytes((byte) 30, readLength));
+
+                    String cigar = "1" + firstOp;
+                    cigar += "1" + middleOp;
+                    cigar += "1" + lastOp;
+                    read.setCigar(cigar);
+
+                    tests.add(new Object[]{read, middleOp});
+                }
+            }
+        }
+
+        return tests.toArray(new Object[][]{});
+    }
+
+    @Test(dataProvider = "PrevAndNextTest_simple")
+    public void testImmediateBeforeAndAfterTest_simple(final GATKRead read, final CigarOperator middleOp) {
```

Comment:
Same comment as above -- redesign to construct `PileupElements` directly rather than confusing things by going through `AlignmentStateMachine`